README.md
# websocket-client([中文文档](./doc/README.zh-cn.md))

## Summary

> A lightweight js lib for websocket.

websocket-client is a lightweight Javascript library running in a browser environment.
Essentially, its instance wrapped around the native WebSocket so that it can provide some commonly used logic(e.g. hearbeat detection).
Give programs similar behaviour to native WebSocket while working to make them more robust. (e.g. registration event, readyState property, etc.) 
The following functions is provided for websocket-client:
- automatic reconnection when an exception closed.
- you can specify the number of automatic reconnections and the interval between connections when connect to the server failed.
- cache sent failed messages and send in cached order at 100 ms intervals when the websocket reconnects.
- hearbeat detection
- [EventHub](#eventhub) mechanism

## Browser Compatibility
Browser versions：
- chrome >= 43
- firefox >= 19
- ie >= 10
- safari >= 10
- edge >= 12
- opera >= 30
- android >= 4.4
- ios >= 10

[comment]:(TODO:)
[comment]:(chrome--16--readyState(not))
[comment]:(firefox--11--readyState(not))
[comment]:(safari--7--readyState(not))
[comment]:(ie--16)
[comment]:(edge--12)
[comment]:(opera--12.1--readyState(not))
[comment]:(ios--6--readyState(not))
[comment]:(https://developer.mozilla.org/en-US/docs/Web/API/WebSocket#Browser_compatibility)

## Install
npm
```Command Line
npm i websocket-client -S
```
or yarn
```Command Line
yarn add websocket-client
```

## Usage

Use in the browser environment like:

```html
<script src="./websocket-client.js"></script>
<script >
    var socket = new WS.WebSocketClient('ws://xxx.com/api/ws', {
        immediateReconnectTimeInterval: 2000,
        heartbeatPeriod: 50000,
        heartbeatMessage: 'ping'
    });
</script>
```

es module:

```javascript
import WebSocketClient from 'websocket-client'
// or
// import { WebSocketClient } from 'websocket-client'

const socket = new WebSocketClient('ws://xxx.com/api/ws', {
    immediateReconnectTimeInterval: 2000,
    heartbeatPeriod: 50000,
    heartbeatMessage: 'ping'
})
socket.onopen = function() {
    console.log('open')
}

socket.onmessage = function(e) {
    console.log('message', e.data)
}

socket.onclose = function() {
    console.log('close')
}

socket.onerror = function(e) {
    console.error(e)
}
```

## API

### new WebSocketClient(url[, options])
- Arguments：

    - `{string} url` The URL to which to connect.
    - `{object} options` see also [options](#options)。
    
- Usage：

    Create a `WebSocketClient` instance and connect to server.

### socketClient.close([code, reason])
- Arguments：

    - `{number} code`
    A numeric value indicating the status code explaining why the connection is being closed.
    Default: `1005`
    - `{string} reason`
    A human-readable string explaining why the connection is closing.
    This string must be no longer than 123 bytes of UTF-8 text (not characters).
    
- Usage：

    Closes the `WebSocket` connection or connection attempt, if any. 
    If the connection is already `CLOSED`, this method does nothing.
    
- Exceptions thrown

    - `INVALID_ACCESS_ERR`
    An invalid `code` was specified.
    - `SYNTAX_ERR`
    The `reason` string is too long or contains unpaired surrogates.

### socketClient.send(data)
- Arguments：

    - {string | Blob | ArrayBuffer | ArrayBufferView | Object} data
    
- Usage：

    Enqueues the specified data to be transmitted to the server over the WebSocket connection,
    increasing the value of `bufferedAmount` by the number of bytes needed to contain the data.
    If the data can't be sent (for example, because it needs to be buffered but the buffer is full),
    the socket is closed automatically.
    
    > Tips: if `data` is a plain object, it will be converted to `JSON` before sending.
    
- Example：
```javascript
var data = { username: 'jhail', age: 18}
socketClient.send(data) // the same as socketClient.send('{"username":"jhail","age":18}')
```

### socketClient.on( event, callback )
- Arguments：

    - {string | Array<string>} event
    - {Function} callback
    
- Usage：

    Listen for a custom event on the current instance(`socketClient`).
    Events can be triggered by `socketClient.trigger`. 
    The callback will receive all the additional arguments passed into these event-triggering methods.
    
    By default, when a message is received,
    the socketClient automatically triggers events corresponding to `message` property values in the received data.
    
    See also [options[message]](#options).
    
- Example：
```javascript
socketClient.on('message', function(msg) {
    console.log(msg)
})
socketClient.trigger('message', 'Hi')
// "Hi"
socketClient.trigger('message', 'Hi')
// "Hi"
```

### socketClient.once( event, callback )
- Arguments：

    - {string} event
    - {Function} callback
    
- Usage：

    Listen for a custom event, but only once. The listener will be removed once it triggers for the first time.
    
- Example：
```javascript
socketClient.once('once', function(msg) {
    console.log(msg)
})
socketClient.trigger('once', 'Hi')
// "Hi"
socketClient.trigger('once', 'Hi') //  If triggered again, no more output "Hi".
```


### socketClient.off( [event, callback] )
- Arguments：

    - {string | Array<string>} event
    - {Function} [callback]
    
- Usage：

    Remove custom event listener(s).
    
    - If no arguments are provided, remove all event listeners;
    - If only the event is provided, remove all listeners for that event;
    - If both event and callback are given, remove the listener for that specific callback only.

### socketClient.trigger( eventName, […args] )
- Arguments：

    - {string} eventName
    - [...args]
    
- Usage：
  
    Trigger an event on the current instance.
    Any additional arguments will be passed into the listener’s callback function.

- Example：
```javascript
socketClient.on('messages', function(a, b) {
    console.log(a + b)
})
socketClient.trigger('messages', 'Hello ', 'World')
// "Hello World"
```


## options

| Attribute                      | Type                                                       | Default   | Description                                                  |
| ------------------------------ | ---------------------------------------------------------- | --------- | ------------------------------------------------------------ |
| connectImmediately             | boolean                                                    | true      | sometimes we don't need to start connecting to the server immediately when we create WebSocketClient instances. When set to `false`, you can establish connections through `socketClient.createWebSocket()` when needed . |
| enableDefaultEventHandler      | boolean                                                    | true      | whether to enable the default event handler function. Default event handling functions are designed to provide heartbeat detection, automatic reconnection when an exception closed, etc.  The default event handler function do not affect the `onopen`、`onmessage`、`onclose`、`onerror` of these events you register. |
| connectLogSilent               | boolean                                                    | false     | connection-related logs.                                     |
| eventLogSilent                 | boolean                                                    | false     | logs when `onopen`、`onclose`、`onerror` event is triggered. |
| maxReconnectTimes              | number                                                     | 3         | the number of attempts to reconnect when the connection fails. |
| reconnectTimeInterval          | number                                                     | 30000     | the time interval before trying to reconnect when the connection fails. |
| immediateReconnectTimeInterval | number                                                     | 2000      | the time interval for trying to reconnect when the connection is abnormally closed. |
| enableMessageCache             | boolean                                                    | false     | whether to enable the message cache mechanism. when set to `true`, it will be cache all sent failed messages and send in cached order at intervals of 100 ms when the websocket reconnects. |
| message                        | string                                                     | 'message' | specify which key of message object is used as the message type. |
| heartbeatPeriod                | number                                                     | 50000     | cycle of heartbeat detection. In milliseconds                |
| heartbeatMessage               | string \| object \| Blob \| ArrayBuffer \| ArrayBufferView | "ping"    | the message of heartbeat detection.                          |
| heartbeatLogSilent             | boolean                                                    | false     | time log of heartbeat detection.                             |

## EventHub

EventHub is designed to provide users with a more convenient and flexible way to handle received messages.
For this purpose, an EventHub API is provided for users to import independently.

### Demo

```javascript
// util.js
import { EventHub } from 'websocket-client'
export const eventHub = new EventHub()

// somewhere.js
import { eventHub } from 'util.js'
eventHub.on('message', ( param ) => {
    console.log(param)
})

// websocket.js
import WebSocketClient from 'websocket-client'
import { eventHub } from 'util'

const socket = new WebSocketClient('ws://xxx.com/api/ws', {
    immediateReconnectTimeInterval: 2000,
    heartbeatPeriod: 50000,
    heartbeatMessage: 'ping'
})

socket.onmessage = function(e) { 
  const data = e.data
  eventHub.trigger(data.message, data)
}
```

### EventHub API

####  eventHub.on(event, callback)
See also [socketClient.on( event, callback )](#socketclienton-event-callback-)

####  eventHub.once( event, callback )
See also [socketClient.once( event, callback )](#socketclientonce-event-callback-)

####  eventHub.off( [event, callback] )
See also [socketClient.off( [event, callback] )](#socketclientoff-event-callback-)

####  eventHub.trigger( eventName, […args] )
See also [socketClient.trigger( eventName, […args] )](#socketclienttrigger-eventname-args-)



CHANGELOG.md
# 1.0.0-alpha.0 (2022-01-26)


### Bug Fixes

* **readme.md:** fixed a comment problem ([8f26936](https://github.com/xiaohaifengke/websocket-client/commit/8f2693600cce5364f15e64196c611f3781d61357))


### Features

* **src:** adjust modules ([8ceaf35](https://github.com/xiaohaifengke/websocket-client/commit/8ceaf3541607949ce62992d142d17137226ba790))



## 0.5.6 (2022-01-25)


### Bug Fixes

* **readme.md:** fixed a comment problem ([8f26936](https://github.com/xiaohaifengke/websocket-client/commit/8f2693600cce5364f15e64196c611f3781d61357))





es/index.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { Heartbeat } from './heartbeat';
import { isPromise, getTime, isArrayBufferView, isPlainObject } from './util';
var defaultOptions = {
  connectImmediately: true,
  enableDefaultEventHandler: true,
  connectLogSilent: false,
  eventLogSilent: false,
  maxReconnectTimes: 3,
  reconnectTimeInterval: 30000,
  immediateReconnectTimeInterval: 2000,
  enableMessageCache: false,
  message: 'message',
  heartbeatPeriod: 50000,
  heartbeatMessage: 'ping',
  heartbeatLogSilent: false
};
export var WebSocketClient = /*#__PURE__*/function () {
  /**
   * WebSocketClient constructor
   * @param { string }  url
   * @param { object | undefined }  options
   */
  function WebSocketClient(url) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, WebSocketClient);

    this.url = url;
    this.options = options;
    this.defaultOptions = _objectSpread({}, defaultOptions);
    this.$options = _objectSpread(_objectSpread({}, this.defaultOptions), options);
    this.wasClean = null; // Boolean. 是否正常断开。一般异常断开时，该值为false

    this.reconnectTimes = 0; // 第n次重连的flag

    this._messageCache = [];
    if (this.$options.connectImmediately) this.createWebSocket();
    this.heartbeat = new Heartbeat(this, {
      message: this.$options.heartbeatMessage,
      period: this.$options.heartbeatPeriod
    });

    this._init();
  }

  _createClass(WebSocketClient, [{
    key: "createWebSocket",
    value: function createWebSocket() {
      if ('WebSocket' in window) {
        if (this.readyState === undefined || this.readyState === WebSocket.CLOSED) {
          this.ws = new WebSocket(this.url);
          if (this.$options.enableDefaultEventHandler) this.initEventsHandler();
        }
      } else {
        console.log('WebSocket is not supported in your browser!');
      }
    }
  }, {
    key: "reconnectWebSocket",
    value: function reconnectWebSocket() {
      var _this = this;

      var times = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
      var interval = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.$options.reconnectTimeInterval;

      if (times < this.$options.maxReconnectTimes) {
        if (!this.$options.connectLogSilent) {
          console.log("Try to reconnect the WebSocket after ".concat(interval / 1000, " seconds."));
        }

        setTimeout(function (t) {
          if (!_this.$options.connectLogSilent) {
            console.log("".concat(t + 1, "th attempt to reconnect to webSocket server."));
          }

          _this.reconnectTimes++;

          _this.createWebSocket();
        }, interval, times);
      } else {
        if (!this.$options.connectLogSilent) {
          console.error("Could not create connection to websocket server. Attempted reconnect ".concat(this.$options.maxReconnectTimes, " times. Giving up!"));
        }
      }
    }
  }, {
    key: "initEventsHandler",
    value: function initEventsHandler() {
      var _this2 = this;

      this.ws.addEventListener('open', function () {
        if (!_this2.$options.eventLogSilent) {
          console.log("The Websocket has opened at ".concat(getTime(), "."));
        }

        _this2.wasClean = null;
        _this2.reconnectTimes = 0;

        _this2.heartbeat.recheck();

        _this2.sendCachedMessage();
      });
      this.ws.addEventListener('message', function (event) {
        _this2.heartbeat.recheck();

        var data = JSON.parse(event.data);

        if (isPlainObject(data)) {
          var message = _this2.$options.message;
          var eventName = data[message];
          eventName && typeof eventName === 'string' && _this2.trigger(eventName, data);
        }
      });
      this.ws.addEventListener('close', function (event) {
        _this2.wasClean = !!event.wasClean;

        if (_this2.wasClean) {
          if (!_this2.$options.eventLogSilent) {
            console.log("The WebSocket connection has been closed normally at ".concat(getTime(), "."), event);
          }
        } else {
          if (!_this2.$options.eventLogSilent) {
            console.error("closed at ".concat(getTime(), "."), event);
          }

          _this2.reconnectWebSocket(_this2.reconnectTimes, _this2.reconnectTimes > 0 ? _this2.$options.reconnectTimeInterval : _this2.$options.immediateReconnectTimeInterval);
        }

        _this2.heartbeat.stopCheck();
      });
      this.ws.addEventListener('error', function (error) {
        if (!_this2.$options.eventLogSilent) {
          console.error("The WebSocket connection has been closed due to an error at ".concat(getTime(), "."), error);
        }
      });
    }
  }, {
    key: "cacheMessage",
    value: function cacheMessage(message) {
      var index = this._messageCache.indexOf(message);

      if (index === -1) {
        this._messageCache.push(message);
      }
    }
  }, {
    key: "sendCachedMessage",
    value: function sendCachedMessage() {
      var _this3 = this;

      if (this._messageCache.length === 0) return;

      var msg = this._messageCache.shift();

      this.send(msg);
      setTimeout(function () {
        _this3.sendCachedMessage();
      }, 100);
    }
    /**
     * 1. 判断当前websocket的连接状态为OPEN状态时发送消息
     * 2. TODO: 当非OPEN状态时，判断是否为正常断开。只有当websocket为异常断开时，才存储信息
     * 3. FIXME: 某些早期浏览器不支持readyState属性，故该方法内的逻辑应该添加相应的判断
     * @param { object } message
     */

  }, {
    key: "send",
    value: function send(message) {
      switch (this.readyState) {
        case undefined:
          {
            /**
             * WebSocket is not instantiated. Store the message to the messageCache
             * and send it after the WebSocket open. Then invoke the createWebSocket method.
             *
             */
            if (message !== this.$options.heartbeatMessage && this.$options.enableMessageCache) {
              this.cacheMessage(message);
            }

            this.createWebSocket();
            break;
          }

        case WebSocket.CONNECTING:
          {
            /**
             * WebSocket is connecting. Store the message to the messageCache
             * and send it after the WebSocket open.
             *
             */
            if (message !== this.$options.heartbeatMessage && this.$options.enableMessageCache) {
              this.cacheMessage(message);
            }

            break;
          }

        case WebSocket.OPEN:
          {
            // WebSocket is connected and send the message directly.
            if (message === this.$options.heartbeatMessage) {
              console.log("Heartbeat check at ".concat(getTime(), "."));
              this.ws.send(JSON.stringify(message));
              break;
            }

            var index = this._messageCache.indexOf(message);

            if (index !== -1) {
              // 如果该msg在messageCache内，则从messageCache中删除该msg
              this._messageCache.splice(index, 1);
            }

            if (message instanceof Blob || message instanceof ArrayBuffer || isArrayBufferView(message)) {
              this.ws.send(message);
            } else {
              this.ws.send(JSON.stringify(message));
            }

            break;
          }

        default:
          {
            /**
             * TODO: Do not store the message if this.wasClean is true
             * Store the message to the messageCache
             * and send it after the WebSocket is reconnected.
             */
            if (message !== this.$options.heartbeatMessage && this.$options.enableMessageCache) {
              this.cacheMessage(message);
            }
          }
      }
    }
    /**
     * TODO:
     * It may be helpful to examine the socket's bufferedAmount attribute before attempting to
     * close the connection to determine if any data has yet to be transmitted on the network.
     * If this value isn't 0, there's pending data still, so you may wish to wait before closing the connection.
     */

  }, {
    key: "close",
    value: function close() {
      var _this$ws;

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      (_this$ws = this.ws) === null || _this$ws === void 0 ? void 0 : _this$ws.close.apply(_this$ws, args);
    }
  }, {
    key: "_init",
    value: function _init() {
      Object.defineProperties(this, {
        readyState: {
          enumerable: true,
          get: function get() {
            var _this$ws2;

            return (_this$ws2 = this.ws) === null || _this$ws2 === void 0 ? void 0 : _this$ws2.readyState;
          }
        },
        status: {
          enumerable: true,
          get: function get() {
            return this.readyState >= 0 ? this.constructor.STATUS[this.readyState] : 'UNCREATE';
          }
        },
        onopen: {
          enumerable: true,
          get: function get() {
            var _this$ws3;

            return (_this$ws3 = this.ws) === null || _this$ws3 === void 0 ? void 0 : _this$ws3.onopen;
          },
          set: function set(fn) {
            this.ws && (this.ws.onopen = fn);
          }
        },
        onmessage: {
          enumerable: true,
          get: function get() {
            var _this$ws4;

            return (_this$ws4 = this.ws) === null || _this$ws4 === void 0 ? void 0 : _this$ws4.onmessage;
          },
          set: function set(fn) {
            this.ws && (this.ws.onmessage = fn);
          }
        },
        onclose: {
          enumerable: true,
          get: function get() {
            var _this$ws5;

            return (_this$ws5 = this.ws) === null || _this$ws5 === void 0 ? void 0 : _this$ws5.onclose;
          },
          set: function set(fn) {
            this.ws && (this.ws.onclose = fn);
          }
        },
        onerror: {
          enumerable: true,
          get: function get() {
            var _this$ws6;

            return (_this$ws6 = this.ws) === null || _this$ws6 === void 0 ? void 0 : _this$ws6.onerror;
          },
          set: function set(fn) {
            this.ws && (this.ws.onerror = fn);
          }
        }
      });
    }
  }, {
    key: "on",
    value: function on(event, func) {
      if (!this._events) this._events = Object.create(null);

      if (Array.isArray(event)) {
        for (var i = 0, l = event.length; i < l; i++) {
          this.on(event[i], func);
        }
      } else {
        ;
        (this._events[event] || (this._events[event] = [])).push(func);
      }

      return this;
    }
  }, {
    key: "once",
    value: function once(event, func) {
      var $this = this;

      function once() {
        $this.off(event, once);
        func.apply($this, arguments);
      }

      once.func = func;
      this.on(event, once);
      return this;
    }
  }, {
    key: "off",
    value: function off(event, func) {
      // all
      if (!arguments.length) {
        this._events = Object.create(null);
        return this;
      } // array of events


      if (Array.isArray(event)) {
        for (var _i = 0, l = event.length; _i < l; _i++) {
          this.off(event[_i], func);
        }

        return this;
      } // specific event


      var cbs = this._events[event];

      if (!cbs) {
        return this;
      }

      if (!func) {
        this._events[event] = null;
        return this;
      } // specific handler


      var cb;
      var i = cbs.length;

      while (i--) {
        cb = cbs[i];

        if (cb === func || cb.func === func) {
          cbs.splice(i, 1);
          break;
        }
      }

      return this;
    }
  }, {
    key: "trigger",
    value: function trigger(event) {
      var cbs = this._events[event];

      if (cbs) {
        cbs = _toConsumableArray(cbs);
        var res;

        for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
          args[_key2 - 1] = arguments[_key2];
        }

        for (var i = 0, l = cbs.length; i < l; i++) {
          try {
            res = cbs[i].apply(this, args);

            if (res && isPromise(res) && !res._handled) {
              res.catch(function (e) {
                return console.error(e);
              });
              res._handled = true;
            }
          } catch (e) {
            console.error(e);
          }
        }
      }

      return this;
    }
  }]);

  return WebSocketClient;
}();

_defineProperty(WebSocketClient, "STATUS", ['CONNECTING', 'OPEN', 'CLOSING', 'CLOSED']);

export { Heartbeat };
export { EventHub } from './util';
export default WebSocketClient;
//# sourceMappingURL=index.js.map